benchmarks/cpp/transformer.cpp

```diff
@@ -0,0 +1,228 @@
+  auto ln0_rstd = at::randn({B * S, 1}, options).to(at::kFloat);
+  auto ln1_mean = at::randn({B * S, 1}, options).to(at::kFloat);
+  auto ln1_rstd = at::randn({B * S, 1}, options).to(at::kFloat);
+  auto mha_linear1 = at::rand({B * S, E}, options).to(at::kFloat);
+  auto mha_linear0 = at::rand({B * S, 3 * E}, options);
+  auto mlp_linear1 = at::rand({B * S, 4 * E}, options);
+
+  std::vector<c10::IValue> at_inputs = {
+      x,
+      grad,
+      shardTensor(mha_w0.view({3, E, E}), 1, mesh, communicator).view({1, 3 * E / D, E}),
+      shardTensor(mha_w1, 1, mesh, communicator).unsqueeze(0),
+      shardTensor(mlp_w0, 0, mesh, communicator).unsqueeze(0),
+      shardTensor(mlp_w1, 1, mesh, communicator).unsqueeze(0),
+      mlp_dropout_mask,
+      mha_dropout_mask,
+      shardTensor(sdpa_output, 1, mesh, communicator).unsqueeze(0),
+      shardTensor(sdpa_logsum_exp, 1, mesh, communicator).unsqueeze(0),
+      sdpa_seed,
+      sdpa_offset,
+      ln1_w,
+      ln1_b,
+      ln1_mean,
+      ln1_rstd,
+      ln0_w,
+      ln0_b,
+      ln0_mean,
+      ln0_rstd,
+      shardTensor(mha_linear0, 1, mesh, communicator).unsqueeze(0),
+      mha_linear1.to(at::kFloat),
+      shardTensor(mlp_linear1, 1, mesh, communicator).unsqueeze(0)
+    };
+
+  DistributedTransformer model = DistributedTransformer(D, B, E, H, S, kDropoutProb, kSdpaProb);
+  auto fec = model.backward(dtype);
+  std::vector<at::Tensor> outputs;
+
+  cudaSetDevice(communicator->deviceId());
+  auto start = std::chrono::high_resolution_clock::now();
+  for (auto i : c10::irange(num_itrs + warmup_itrs)) {
+    if (i == warmup_itrs) {
+      start = std::chrono::high_resolution_clock::now();
+    }
+    if (i >= warmup_itrs && profile) {
+      nvtxRangePush(("BwdIteration" + std::to_string(i)).c_str());
+    }
+    outputs = fec->runFusionWithInputs(at_inputs);
+    cudaDeviceSynchronize();
+    // cudaDeviceSynchronize is not blocking until kernels are finished on all
+    // devices except 0
+    // TODO: are we not waiting until all kernels are appended to the stream?
+    //std::cout << outputs[0][0][0][0] << std::endl;
+
+    if (i >= warmup_itrs && profile) {
+      nvtxRangePop();
+    }
+  }
+  auto end = std::chrono::high_resolution_clock::now();
+  if (profile) {
+    cudaProfilerStop();
+  }
+
+  double backward_time =
+      std::chrono::duration_cast<std::chrono::microseconds>(end - start)
+          .count() /
+      (double)num_itrs / 1000.0;
+  std::cout << communicator->deviceId() << ": Average backward time "
+            << backward_time << "ms" << std::endl;
+}
+
+int main(int argc, char** argv) {
+  bool profile = false;
+  bool sequence_parallel = false;
+  if (argc == 3) {
+    profile = (bool)atoi(argv[1]);
+    sequence_parallel = (bool)atoi(argv[2]);
+  }
+  auto communicator = &Communicator::getInstance();
+  forward_transformer(communicator, profile, sequence_parallel);
+  communicator->barrier();
```

Comment:
you probably want: 
```suggestion
  communicator->barrier()->wait();
```
right?